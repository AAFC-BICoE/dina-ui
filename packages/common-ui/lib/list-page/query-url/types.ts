export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
