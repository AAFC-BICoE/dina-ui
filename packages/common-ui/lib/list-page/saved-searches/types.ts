import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
