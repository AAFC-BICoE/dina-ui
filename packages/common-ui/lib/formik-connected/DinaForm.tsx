import {
  Form,
  Formik,
  FormikConfig,
  FormikConsumer,
  FormikContextType,
  FormikProps,
  FormikValues
} from "formik";
import { cloneDeep } from "lodash";
import { FormTemplate } from "../../../dina-ui/types/collection-api";
import {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useMemo
} from "react";
import { useIntl } from "react-intl";
import { BulkEditTabContext, scrollToError } from "..";
import { AccountContextI, useAccount } from "../account/AccountProvider";
import { ApiClientI, useApiClient } from "../api-client/ApiClientContext";
import { ErrorViewer } from "./ErrorViewer";
import { safeSubmit } from "./safeSubmit";

export interface DinaFormProps<TValues>
  extends Omit<FormikConfig<TValues>, "onSubmit">,
    Omit<DinaFormContextI, "initialValues"> {
  onSubmit?: DinaFormOnSubmit<TValues>;
}

/** Values available to form elements. */
export interface DinaFormContextI {
  readOnly?: boolean;
  /*
   * Whether to layout the label and value horizontally (Default vertical).
   * If a [number, number] is passed then those are the bootstrap grid columns for the label and value.
   * "true" defaults to [6, 6].
   */
  horizontal?: boolean | [number, number] | "flex";

  /** The initial form values passed into Formik. */
  initialValues?: any;

  /** Add a checkbox beside the wrapper field if true */
  isTemplate?: boolean;

  /**
   * Form template with all restrictions to place on the form.
   */
  formTemplate?: FormTemplate;

  /**
   * The component name for all of the fields within this dina form. Using DinaFormContext you can
   * override it.
   */
  componentName?: string;

  /**
   * The section name for all of the fields within this dina form. Using the DinaFormContext you can
   * override it.
   */
  sectionName?: string;

  /**
   * @deperecated
   * Whether this DinaForm is nested in another DinaForm. Nested forms are bad so avoid this.
   */
  isNestedForm?: boolean;
}

export type DinaFormOnSubmit<TValues = any> = (
  params: DinaFormSubmitParams<TValues>
) => Promise<void> | void;

export interface DinaFormSubmitParams<TValues> {
  submittedValues: TValues;
  formik: FormikContextType<TValues>;
  api: ApiClientI;
  account: AccountContextI;
}

/** Wrapps Formik with safe error handling+displaying and API/Account onSubmit params. */
export function DinaForm<Values extends FormikValues = FormikValues>(
  props: DinaFormProps<Values>
) {
  const api = useApiClient();
  const account = useAccount();

  const isNestedForm = !!useContext(DinaFormContext);

  const { children: childrenProp, onSubmit: onSubmitProp, readOnly } = props;

  /** Wrapped onSubmit prop with erorr handling and API/Account params. */
  const onSubmitInternal = safeSubmit(async (submittedValues, formik) => {
    // Make a copy of the submitted values so the original can't be mutated in the passed onSubmit function:
    const submittedValuesCopy = cloneDeep(submittedValues);
    try {
      await onSubmitProp?.({
        submittedValues: submittedValuesCopy,
        formik,
        api,
        account
      });
    } catch (error) {
      scrollToError();
      throw error;
    }
  });

  const childrenInternal:
    | ((formikProps: FormikProps<Values>) => React.ReactNode)
    | React.ReactNode =
    typeof childrenProp === "function" ? (
      (formikProps) => <FormWrapper>{childrenProp(formikProps)}</FormWrapper>
    ) : (
      <FormWrapper>{childrenProp}</FormWrapper>
    );

  // Clone the initialValues object so it isn't modified in the form:
  const initialValues = useMemo(
    () => cloneDeep(props.initialValues),
    [props.initialValues]
  );

  /**
   * Disable the bulk edit tab context for nested forms.
   * e.g. Don't show the has-bulk-edit-value indicators in the Material Sample
   * form's nested Collecting Event form.
   */
  const withBulkEditCtx = useCallback<(content: JSX.Element) => JSX.Element>(
    isNestedForm
      ? (content) => (
          <BulkEditTabContext.Provider value={null}>
            {content}
          </BulkEditTabContext.Provider>
        )
      : (content) => content,
    [isNestedForm]
  );

  return withBulkEditCtx(
    <DinaFormContext.Provider
      value={{
        ...props,
        isNestedForm,
        readOnly: props.readOnly ?? false
      }}
    >
      <Formik
        // Don't use Formik's default validation triggers:
        // Only validate on submit. And remove field error on field value change.
        validateOnChange={false}
        validateOnBlur={false}
        {...props}
        initialValues={initialValues}
        onSubmit={onSubmitInternal}
      >
        {childrenInternal}
      </Formik>
    </DinaFormContext.Provider>
  );
}

/** Wraps the inner content with the Form + ErrorViewer components. */
function FormWrapper({ children }: PropsWithChildren<{}>) {
  const { isNestedForm } = useDinaFormContext();

  // Disable enter to submit form in nested forms.
  function disableEnterToSubmitOuterForm(e) {
    // Pressing enter should not submit the outer form:
    if (e.keyCode === 13 && e.target.tagName !== "TEXTAREA") {
      e.preventDefault();
    }
  }

  const PromptIfDirty = ({ formik }) => {
    const { formatMessage } = useIntl();
    // only prompt if there is data change in edit or add pages
    if (formik.dirty && formik.values.type && formik.submitCount === 0) {
      window.onbeforeunload = () => {
        return formatMessage({ id: "possibleDataLossWarning" });
      };
    } else window.onbeforeunload = null;
    return null;
  };
  const Wrapper = isNestedForm ? "div" : Form;

  return (
    <Wrapper
      onKeyDown={isNestedForm ? disableEnterToSubmitOuterForm : undefined}
    >
      <ErrorViewer />
      <FormikConsumer>
        {(formik) => (
          <>
            <PromptIfDirty formik={formik} />
          </>
        )}
      </FormikConsumer>
      {children}
    </Wrapper>
  );
}

export const DinaFormContext = createContext<DinaFormContextI | null>(null);

export function useDinaFormContext() {
  const ctx = useContext(DinaFormContext);
  if (!ctx) {
    throw new Error("No DinaFormContext available.");
  }
  return ctx;
}

export type DinaFormSectionProps = PropsWithChildren<Partial<DinaFormContextI>>;

/**
 * Override context values for a section of the form.
 * e.g. making part of the form layout horizontal or readOnly.
 */
export function DinaFormSection({
  children,
  ...ctxOverride
}: DinaFormSectionProps) {
  const ctx = useContext(DinaFormContext);

  return (
    <DinaFormContext.Provider value={{ ...ctx, ...ctxOverride }}>
      {children}
    </DinaFormContext.Provider>
  );
}
