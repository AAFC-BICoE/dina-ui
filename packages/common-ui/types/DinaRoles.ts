export const DINA_ADMIN = "dina-admin";
export const READ_ONLY_ADMIN = "read-only-admin";
export const SUPER_USER = "super-user";
export const USER = "user";
export const GUEST = "guest";
export const READ_ONLY = "read-only";
