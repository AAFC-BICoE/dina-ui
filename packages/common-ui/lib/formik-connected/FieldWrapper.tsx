import classNames from "classnames";
import { FastField, FormikProps } from "formik";
import { isArray } from "lodash";
import { ReactNode, useMemo } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { FieldHeader } from "../field-header/FieldHeader";
import { CheckBoxWithoutWrapper } from "./CheckBoxWithoutWrapper";
import { useDinaFormContext } from "./DinaForm";
import { ReadOnlyValue } from "./FieldView";

export interface LabelWrapperParams {
  /** The CSS classes of the div wrapper. */
  className?: string;

  /** Hides the label. */
  hideLabel?: boolean;

  /** The name of the field. */
  name: string;

  /** The label for the field. */
  label?: ReactNode;

  /** Override the default "name" prop used to get labels and tooltips from the intl messages. */
  customName?: string;

  /** Link href to render for a single string value. */
  link?: string;

  /** Custom element to render when the form is in read-only mode. */
  readOnlyRender?: (value: any, form: FormikProps<any>) => ReactNode;

  removeBottomMargin?: boolean;

  /** Remove the label. */
  removeLabel?: boolean;

  /** Disables how clicking a label clicks the inner element. */
  disableLabelClick?: boolean;

  /** Add an image inside of the tooltip. Provide the URL of the image to display it. */
  tooltipImage?: string;

  /** Accessability text, only used if a tooltip image is provided. */
  tooltipImageAlt?: string;

  /** Add a link to a tooltip. */
  tooltipLink?: string;

  /** The text that appears for the link. */
  tooltipLinkText?: string;

  /**
   * Custom field name for the template checkbox.
   * e.g. passing "srcAdminLevels[0]" will change the default
   * "templateCheckboxes['srcAdminLevels[0].name']"
   * to "templateCheckboxes['srcAdminLevels[0]']".
   */
  templateCheckboxFieldName?: string;

  /** Override FastField's default shouldComponentUpdate */
  shouldUpdate?: (nextProps: LabelWrapperParams, props: {}) => boolean;

  validate?: (value: any) => string | void;
}

export interface FieldWrapperProps extends LabelWrapperParams {
  children?:
    | JSX.Element
    | ((renderProps: FieldWrapperRenderProps) => JSX.Element);
  removeLabelTag?: boolean;
}

export interface FieldWrapperRenderProps {
  invalid: boolean;
  value: any;
  setValue: (newValue: any) => void;
  formik: FormikProps<any>;
}

/**
 * Wraps a field with a label of the field's name and Formik's FastField. The label can be auto-generated as a title-case
 * version of the field name, or can be specified as a custom label string.
 *
 * This component also wraps the field in a div with the className `${fieldName}-field` for testing purposes.
 * e.g. select the "description" text input using wrapper.find(".description-field input").
 */
export function FieldWrapper({
  className,
  disableLabelClick,
  hideLabel = false,
  name,
  label,
  children,
  customName,
  link,
  readOnlyRender,
  removeBottomMargin,
  removeLabel,
  tooltipImage,
  tooltipImageAlt,
  tooltipLink,
  tooltipLinkText,
  templateCheckboxFieldName,
  removeLabelTag,
  shouldUpdate,
  validate
}: FieldWrapperProps) {
  const { horizontal, readOnly, isTemplate, enabledFields } =
    useDinaFormContext();

  /** Whether this field should be hidden because the template doesn't specify that it should be shown. */
  const disabledByFormTemplate = useMemo(
    () =>
      enabledFields
        ? !enabledFields.includes(templateCheckboxFieldName || name)
        : false,
    [enabledFields]
  );

  const fieldLabel = label ?? (
    <FieldHeader
      name={name}
      customName={customName}
      tooltipImage={tooltipImage}
      tooltipImageAlt={tooltipImageAlt}
      tooltipLink={tooltipLink}
      tooltipLinkText={tooltipLinkText}
    />
  );

  const [labelClass, valueClass] =
    horizontal === true
      ? ["col-sm-6", "col-sm-6"]
      : horizontal === "flex"
      ? ["", "flex-grow-1"]
      : (horizontal || []).map(col => `col-sm-${col}`) ||
        (isTemplate ? ["col-sm-12", "col-sm-12"] : []);

  if (disabledByFormTemplate) {
    return null;
  }

  const fieldWrapperInternal = (
    <div className={valueClass} style={{ cursor: "auto" }}>
      <FastField name={name} shouldUpdate={shouldUpdate} validate={validate}>
        {({ field: { value }, form, meta: { error } }) => (
          <ErrorBoundary
            // The error boundary is just for render errors
            // so an error thrown in a form field's render function kills just that field,
            // not the whole page.
            FallbackComponent={({ error: renderError }) => (
              <div className="alert alert-danger" role="alert">
                <pre className="mb-0">{renderError.message}</pre>
              </div>
            )}
          >
            {readOnly || !children
              ? readOnlyRender?.(value, form) ?? (
                  <ReadOnlyValue link={link} value={value} />
                )
              : typeof children === "function"
              ? children?.({
                  invalid: Boolean(error),
                  value,
                  setValue: newValue => {
                    // Remove the error message when the user edits the field:
                    form.setFieldError(name, undefined);
                    form.setFieldValue(name, newValue);
                    form.setFieldTouched(name);
                  },
                  formik: form
                })
              : children}
            {error && <div className="invalid-feedback">{error}</div>}
          </ErrorBoundary>
        )}
      </FastField>
    </div>
  );

  return (
    <div className={classNames(className, { row: isTemplate })}>
      {isTemplate && (
        <CheckBoxWithoutWrapper
          name={`templateCheckboxes['${templateCheckboxFieldName ?? name}']`}
          className="col-sm-1 templateCheckBox"
        />
      )}
      {removeLabelTag ? (
        <>
          {!removeLabel && (
            <div className={classNames(labelClass, !horizontal && "mb-2")}>
              {!hideLabel && <strong>{fieldLabel}</strong>}
            </div>
          )}

          {fieldWrapperInternal}
        </>
      ) : (
        <label
          className={classNames(
            `${name}-field`,
            customName && `${customName}-field`,
            horizontal === "flex" && "d-flex gap-2",
            horizontal ? "align-items-center" : "mb-2",
            (horizontal === true || isArray(horizontal)) && "row",
            isTemplate && `col-sm-${horizontal ? "11" : "10"}`,
            !isTemplate && !horizontal && "w-100",
            !removeBottomMargin && "mb-3"
          )}
          htmlFor={disableLabelClick ? "none" : undefined}
        >
          {!removeLabel && (
            <div className={classNames(labelClass, !horizontal && "mb-2")}>
              {!hideLabel && <strong>{fieldLabel}</strong>}
            </div>
          )}
          {fieldWrapperInternal}
        </label>
      )}
    </div>
  );
}
