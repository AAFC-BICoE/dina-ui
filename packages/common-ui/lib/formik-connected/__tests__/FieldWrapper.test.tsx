import { FormTemplate } from "../../../../dina-ui/types/collection-api";
import { IntlProvider } from "react-intl";
import { mountWithAppContext } from "../../test-util/mock-app-context";
import { DinaForm, DinaFormSection } from "../DinaForm";
import { FieldWrapper } from "../FieldWrapper";
import { BulkEditContext, BulkEditTabContext } from "../..";

const mockSubmit = jest.fn();

describe("FieldWrapper component.", () => {
  beforeEach(jest.clearAllMocks);

  describe("Field wrapper display props", () => {
    it("Adds a generated title-case label to the wrapped component.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <FieldWrapper name="fieldName">{() => <div />}</FieldWrapper>
        </DinaForm>
      );

      expect(wrapper.find("label").text()).toEqual("Field Name");
    });

    it("Accepts a className which is applied to a surrounding div.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <FieldWrapper className="col-6" name="fieldName">
            {() => <div />}
          </FieldWrapper>
        </DinaForm>
      );

      expect(wrapper.find(".col-6").exists()).toEqual(true);
    });

    it("Displays the intl message (if there is one) in the label.", () => {
      const wrapper = mountWithAppContext(
        <IntlProvider
          locale="en"
          messages={{ field_testField: "My Field Label" }}
        >
          <DinaForm initialValues={{}}>
            <FieldWrapper name="testField">{() => <div />}</FieldWrapper>
          </DinaForm>
        </IntlProvider>
      );

      expect(wrapper.find("label").text()).toEqual("My Field Label");
    });

    it("Displays a custom label.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <FieldWrapper label="The Group's Name" name="group.groupName">
            {() => <div />}
          </FieldWrapper>
        </DinaForm>
      );

      // Snapshot ensures the structure does not change.
      expect(wrapper.find(FieldWrapper).html()).toMatchSnapshot();
    });

    it("Displays horizontal custom label.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}} horizontal={true}>
          <FieldWrapper label="The Group's Name" name="group.groupName">
            {() => <div />}
          </FieldWrapper>
        </DinaForm>
      );

      // Snapshot ensures the structure does not change.
      expect(wrapper.find(FieldWrapper).html()).toMatchSnapshot();
    });
  });

  describe("Read only views", () => {
    it("Displays the readOnly value when the form is read-only.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{ myField: "my value" }} readOnly={true}>
          <FieldWrapper name="myField" />
        </DinaForm>
      );

      expect(wrapper.find(".field-view").text()).toEqual("my value");
    });

    it("Can display a custom read-only view.", () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{ myField: "my value" }} readOnly={true}>
          <FieldWrapper
            name="myField"
            readOnlyRender={(value) => (
              <div className="custom-div">{value}</div>
            )}
          />
        </DinaForm>
      );

      expect(wrapper.find(".custom-div").text()).toEqual("my value");
    });
  });

  describe("Form Template functionality", () => {
    const FORM_TEMPLATE_DEFAULT_VALUES: FormTemplate = {
      type: "form-template",
      components: [
        {
          name: "testComponent",
          visible: true,
          order: 0,
          sections: [
            {
              name: "testSection1",
              visible: true,
              items: [
                {
                  name: "testField1",
                  visible: true,
                  defaultValue: 1
                },
                {
                  name: "testField2",
                  visible: true,
                  defaultValue: true
                }
              ]
            },
            {
              name: "testSection2",
              visible: true,
              items: [
                {
                  name: "testField3",
                  visible: true,
                  defaultValue: "defaultValue1"
                },
                {
                  name: "testField4",
                  visible: true,
                  defaultValue: "defaultValue2"
                }
              ]
            }
          ]
        }
      ]
    };

    const DINA_FORM_DEFAULT_VALUES = (disableSection) => (
      <DinaFormSection componentName="testComponent">
        <DinaFormSection sectionName="testSection1">
          <FieldWrapper name="testField1">
            {(props) => <div className="testField1">{props.value}</div>}
          </FieldWrapper>
          <FieldWrapper name="testField2">
            {(props) => (
              <div className="testField2">{props.value ? "true" : "false"}</div>
            )}
          </FieldWrapper>
        </DinaFormSection>

        <DinaFormSection sectionName="testSection2">
          <FieldWrapper
            name="testField3"
            disableFormTemplateDefaultValue={disableSection}
          >
            {(props) => <div className="testField3">{props.value}</div>}
          </FieldWrapper>
          <FieldWrapper
            name="testField4"
            disableFormTemplateDefaultValue={disableSection}
          >
            {(props) => <div className="testField4">{props.value}</div>}
          </FieldWrapper>
        </DinaFormSection>
      </DinaFormSection>
    );

    it("Accepts a custom field name for the template checkbox.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          initialValues={{ myField: "my value", templateCheckboxes: {} }}
          isTemplate={true}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
        >
          <DinaFormSection
            componentName="componentName"
            sectionName="sectionName"
          >
            <FieldWrapper
              templateCheckboxFieldName="customTemplateFieldName"
              name="myField"
            >
              {({ value }) => <>{value}</>}
            </FieldWrapper>
          </DinaFormSection>
        </DinaForm>
      );

      wrapper
        .find("input[type='checkbox']")
        .simulate("change", { target: { checked: true } });
      wrapper.update();

      wrapper.find("form").simulate("submit");

      await new Promise(setImmediate);
      wrapper.update();

      expect(mockSubmit).lastCalledWith({
        myField: "my value",
        templateCheckboxes: {
          "componentName.sectionName.customTemplateFieldName": true
        }
      });
    });

    it("Properly hides the field when the form template has disabled it.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          initialValues={{ myField1: "my value", templateCheckboxes: {} }}
          formTemplate={{
            type: "form-template",
            components: [
              {
                name: "testComponent",
                visible: true,
                order: 0,
                sections: [
                  {
                    name: "enabledSection",
                    visible: true,
                    items: [
                      {
                        name: "enabledField1",
                        visible: true
                      },
                      {
                        name: "customTemplateFieldName",
                        visible: true
                      }
                    ]
                  },
                  {
                    name: "disabledSection",
                    visible: true,
                    items: [
                      {
                        name: "disabledField1",
                        visible: false
                      },
                      {
                        name: "disabledCustomTemplateFieldName",
                        visible: false
                      }
                    ]
                  }
                ]
              }
            ]
          }}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
        >
          <DinaFormSection componentName="testComponent">
            <DinaFormSection sectionName="enabledSection">
              {/* Enabled Fields: */}
              <FieldWrapper name="enabledField1">
                {() => <div className="enabledField1" />}
              </FieldWrapper>
              <FieldWrapper
                templateCheckboxFieldName="customTemplateFieldName"
                name="enabledField2"
              >
                {() => <div className="enabledField2" />}
              </FieldWrapper>
            </DinaFormSection>

            <DinaFormSection sectionName="disabledSection">
              {/* Disabled Fields: */}
              <FieldWrapper name="disabledField1">
                {() => <div className="disabledField1" />}
              </FieldWrapper>
              <FieldWrapper
                templateCheckboxFieldName="disabledCustomTemplateFieldName"
                name="disabledField2"
              >
                {() => <div className="disabledField2" />}
              </FieldWrapper>
            </DinaFormSection>
          </DinaFormSection>
        </DinaForm>
      );

      expect(wrapper.find(".enabledField1").exists()).toEqual(true);
      expect(wrapper.find(".enabledField2").exists()).toEqual(true);
      expect(wrapper.find(".disabledField1").exists()).toEqual(false);
      expect(wrapper.find(".disabledField2").exists()).toEqual(false);
    });

    it("Default values are provided from the form template", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          // These values should be overridden.
          initialValues={{
            testField1: 2,
            testField2: false,
            testField3: "initialValue1",
            testField4: "initialValue2"
          }}
          formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
          isExistingRecord={false}
        >
          {DINA_FORM_DEFAULT_VALUES(false)}
        </DinaForm>
      );

      // Expect the form template default values. NOT the form initial values.
      expect(wrapper.find(".testField1").text()).toEqual("1");
      expect(wrapper.find(".testField2").text()).toEqual("true");
      expect(wrapper.find(".testField3").text()).toEqual("defaultValue1");
      expect(wrapper.find(".testField4").text()).toEqual("defaultValue2");

      // Changing the form template to undefined should re-render the form with the forms initial values.
      wrapper.setProps({
        children: (
          <DinaForm
            // These values should be overridden.
            initialValues={{
              testField1: 2,
              testField2: false,
              testField3: "initialValue1",
              testField4: "initialValue2"
            }}
            formTemplate={undefined}
            onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
            isExistingRecord={false}
          >
            {DINA_FORM_DEFAULT_VALUES(false)}
          </DinaForm>
        )
      });
      wrapper.update();

      // Now it should be the forms initial values.
      expect(wrapper.prop("formTemplate")).toBeUndefined();
      expect(wrapper.find(".testField1").text()).toEqual("2");
      expect(wrapper.find(".testField2").text()).toEqual("false");
      expect(wrapper.find(".testField3").text()).toEqual("initialValue1");
      expect(wrapper.find(".testField4").text()).toEqual("initialValue2");
    });

    it("Form template default values should only work for creating new records", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          // These values should be overridden.
          initialValues={{
            testField1: 2,
            testField2: false,
            testField3: "initialValue1",
            testField4: "initialValue2"
          }}
          formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
          isExistingRecord={true} // Editing an existing record.
        >
          {DINA_FORM_DEFAULT_VALUES(false)}
        </DinaForm>
      );

      // The initial values should be used in this case since it's a existing record.
      expect(wrapper.find(".testField1").text()).toEqual("2");
      expect(wrapper.find(".testField2").text()).toEqual("false");
      expect(wrapper.find(".testField3").text()).toEqual("initialValue1");
      expect(wrapper.find(".testField4").text()).toEqual("initialValue2");
    });

    it("Custom default value handling support for fields", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          // These values should be overridden.
          initialValues={{
            testField2: false
          }}
          formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
          isExistingRecord={false}
        >
          <DinaFormSection componentName="testComponent">
            <DinaFormSection sectionName="testSection1">
              <FieldWrapper
                name="testField2"
                customHandleDefaultValue={({
                  formikContext,
                  formTemplate,
                  formTemplateItem,
                  setDefaultValue,
                  initialValues
                }) => {
                  if (formTemplateItem?.defaultValue === true) {
                    setDefaultValue("custom true value");
                  } else {
                    setDefaultValue("custom false value");
                  }

                  // Test using the formik context to update other fields.
                  formikContext.setFieldValue(
                    "testField3",
                    JSON.stringify(formTemplate)
                  );
                  formikContext.setFieldValue(
                    "testField4",
                    JSON.stringify(initialValues)
                  );
                }}
              >
                {(props) => <div className="testField2">{props.value}</div>}
              </FieldWrapper>
              <FieldWrapper
                name="testField3"
                disableFormTemplateDefaultValue={true}
              >
                {(props) => <div className="testField3">{props.value}</div>}
              </FieldWrapper>
              <FieldWrapper
                name="testField4"
                disableFormTemplateDefaultValue={true}
              >
                {(props) => <div className="testField4">{props.value}</div>}
              </FieldWrapper>
            </DinaFormSection>
          </DinaFormSection>
        </DinaForm>
      );

      // Check to see if the custom value is returned.
      expect(wrapper.find(".testField2").text()).toEqual("custom true value");

      // This snapshot contains the form template, to ensure it's being returned.
      expect(wrapper.find(".testField3").html()).toMatchSnapshot();

      // This snapshot contains the initial values, to ensur eit's being returned.
      expect(wrapper.find(".testField4").html()).toMatchSnapshot();
    });

    it("Disable the default values being used from the form template", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm
          // These values should be overridden.
          initialValues={{
            testField1: 2,
            testField2: false,
            testField3: "initialValue1",
            testField4: "initialValue2"
          }}
          formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
          onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
          isExistingRecord={false}
        >
          {DINA_FORM_DEFAULT_VALUES(true)}
        </DinaForm>
      );

      // Expect for field 1 and 2 to be using form template, other fields are disabled.
      expect(wrapper.find(".testField1").text()).toEqual("1");
      expect(wrapper.find(".testField2").text()).toEqual("true");
      expect(wrapper.find(".testField3").text()).toEqual("initialValue1"); // Disabled, use initial.
      expect(wrapper.find(".testField4").text()).toEqual("initialValue2"); // Disabled, use initial.
    });

    it("Bulk editing using form template default values (Edit All Tab)", async () => {
      // Default values should work when editing in the EDIT ALL tab.
      const wrapper = mountWithAppContext(
        <BulkEditContext.Provider value={true}>
          <BulkEditTabContext.Provider
            value={{ bulkEditFormRef: jest.fn as any, resourceHooks: [] }}
          >
            <DinaForm
              // These values should be overridden.
              initialValues={{
                testField1: 2,
                testField2: false,
                testField3: "initialValue1",
                testField4: "initialValue2"
              }}
              formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
              onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
              isExistingRecord={false}
            >
              {DINA_FORM_DEFAULT_VALUES(false)}
            </DinaForm>
          </BulkEditTabContext.Provider>
        </BulkEditContext.Provider>
      );

      // Expect the form template default values. NOT the form initial values.
      expect(wrapper.find(".testField1").text()).toEqual("1");
      expect(wrapper.find(".testField2").text()).toEqual("true");
      expect(wrapper.find(".testField3").text()).toEqual("defaultValue1");
      expect(wrapper.find(".testField4").text()).toEqual("defaultValue2");
    });

    it("Bulk editing using form template default values (Individual Edit Tab)", async () => {
      // Try it again but this time as a individual tab.
      const wrapper = mountWithAppContext(
        <BulkEditContext.Provider value={true}>
          <DinaForm
            // These values should be overridden.
            initialValues={{
              testField1: 2,
              testField2: false,
              testField3: "initialValue1",
              testField4: "initialValue2"
            }}
            formTemplate={FORM_TEMPLATE_DEFAULT_VALUES}
            onSubmit={({ submittedValues }) => mockSubmit(submittedValues)}
            isExistingRecord={false}
          >
            {DINA_FORM_DEFAULT_VALUES(false)}
          </DinaForm>
        </BulkEditContext.Provider>
      );

      // Expect the initial values.
      expect(wrapper.find(".testField1").text()).toEqual("2");
      expect(wrapper.find(".testField2").text()).toEqual("false");
      expect(wrapper.find(".testField3").text()).toEqual("initialValue1");
      expect(wrapper.find(".testField4").text()).toEqual("initialValue2");
    });
  });
});
