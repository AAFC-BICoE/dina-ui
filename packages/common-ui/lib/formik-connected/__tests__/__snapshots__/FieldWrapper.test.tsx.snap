// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper component. Field wrapper display props Displays a custom label. 1`] = `"<div class=\\"\\"><label class=\\"group_groupName-field mb-2 w-100 mb-3\\"><div class=\\"field-label mb-2\\"><strong>The Group's Name</strong></div><div style=\\"cursor: auto;\\"><div></div></div></label></div>"`;

exports[`FieldWrapper component. Field wrapper display props Displays horizontal custom label. 1`] = `"<div class=\\"\\"><label class=\\"group_groupName-field align-items-center row mb-3\\"><div class=\\"field-label col-sm-6\\"><strong>The Group's Name</strong></div><div class=\\"col-sm-6\\" style=\\"cursor: auto;\\"><div></div></div></label></div>"`;

exports[`FieldWrapper component. Form Template functionality Custom default value handling support for fields 1`] = `"<div class=\\"testField3\\">{\\"type\\":\\"form-template\\",\\"components\\":[{\\"name\\":\\"testComponent\\",\\"visible\\":true,\\"order\\":0,\\"sections\\":[{\\"name\\":\\"testSection1\\",\\"visible\\":true,\\"items\\":[{\\"name\\":\\"testField1\\",\\"visible\\":true,\\"defaultValue\\":1},{\\"name\\":\\"testField2\\",\\"visible\\":true,\\"defaultValue\\":true}]},{\\"name\\":\\"testSection2\\",\\"visible\\":true,\\"items\\":[{\\"name\\":\\"testField3\\",\\"visible\\":true,\\"defaultValue\\":\\"defaultValue1\\"},{\\"name\\":\\"testField4\\",\\"visible\\":true,\\"defaultValue\\":\\"defaultValue2\\"}]}]}]}</div>"`;

exports[`FieldWrapper component. Form Template functionality Custom default value handling support for fields 2`] = `"<div class=\\"testField4\\">{\\"testField2\\":false}</div>"`;
